### PR TITLE
reimplement std.traits.VariadicFunctionStyle()

### PR DESCRIPTION
**Pull request by WalterBright**
using __traits
fix Issue 17421 - add __traits(getFunctionVariadicStyle, f)

### CODE REVIEW

**Comment by WalterBright** [comment]
Blocked by https://github.com/dlang/dmd/pull/6825

File: std/traits.d [hunk 1]
```diff
@@ -1911,21 +1916,12 @@ enum Variadic
...
-    // TypeFuncion --> CallConvention FuncAttrs Arguments ArgClose Type
-    enum callconv = functionLinkage!Func;
-    enum mfunc = mangledName!Func;
-    enum mtype = mangledName!(ReturnType!Func);
-    static assert(mfunc[$ - mtype.length .. $] == mtype, mfunc ~ "|" ~ mtype);
-
-    enum argclose = mfunc[$ - mtype.length - 1];
-    static assert(argclose >= 'X' && argclose <= 'Z');
-
+    enum string varargs = __traits(getFunctionVariadicStyle, FunctionTypeOf!func);
```
Comment (wilzbach) [inline std/traits.d]: If you remove the `Unqual`, wouldn't it be good to add unittests for functions with qualifiers?

**Comment by rainers** [comment]
The jenkins failure seems to be from before the compiler update, so I guess we can ignore it here.

File: std/traits.d (hunk 1 above)
Comment (rainers) [inline std/traits.d]: > If you remove Unqual, wouldn't it be good to add unittests for functions with qualifiers?

Qualifiers are pretty orthogonal in the implementation of traits in the compiler, so it's very unlikely to cause trouble. The same pattern is already used in functionLinkage.

File: std/traits.d [hunk 2]
```diff
@@ -1911,21 +1916,12 @@ enum Variadic
...
     enum Variadic variadicFunctionStyle =
-        argclose == 'X' ? Variadic.typesafe :
-        argclose == 'Y' ? (callconv == "C") ? Variadic.c : Variadic.d :
-        Variadic.no; // 'Z'
+        (varargs == "stdarg") ? Variadic.c :
+        (varargs == "argptr") ? Variadic.d :
+        (varargs == "typesafe") ? Variadic.typesafe :
+        (varargs == "none") ? Variadic.no : Variadic.no;
```
Comment (andralex) [inline std/traits.d]: looks redundant, not a blocker

**Pull request merged**